File: apps/ts-app/src/greeter.test.ts
import { greet } from "./greeter";

function assert(condition: boolean, msg: string): void {
  if (!condition) throw new Error(`Assertion failed: ${msg}`);
}

assert(greet("Bazel") === "Hello from TypeScript, Bazel!", "greet Bazel");
assert(greet("World") === "Hello from TypeScript, World!", "greet World");

console.log("All TypeScript tests passed.");

File: apps/java-app/src/test/java/com/example/GreeterTest.java
package com.example;

// No external test framework — uses only AssertionError from the JDK.
// This keeps the build fully hermetic (no rules_jvm_external / local JRE needed).
// Bazel downloads the remote JDK via --java_runtime_version=remotejdk_21.
public class GreeterTest {
    public static void main(String[] args) {
        testGreet();
        testGreetWorld();
        System.out.println("All Java tests passed.");
    }

    static void testGreet() {
        Greeter g = new Greeter("Bazel");
        String result = g.greet();
        assert result.equals("Hello from Java, Bazel!") :
            "Expected 'Hello from Java, Bazel!' but got: " + result;
    }

    static void testGreetWorld() {
        Greeter g = new Greeter("World");
        String result = g.greet();
        assert result.equals("Hello from Java, World!") :
            "Expected 'Hello from Java, World!' but got: " + result;
    }
}
